linux/src/threadpool_linux.c

```diff
@@ -955,33 +857,21 @@ void threadpool_destroy_work_item(THANDLE(THREADPOOL) threadpool, THREADPOOL_WOR
     else
     {
         THREADPOOL* threadpool_ptr = THANDLE_GET_T(THREADPOOL)(threadpool);
-
-        /* Codes_SRS_THREADPOOL_LINUX_05_031: [ threadpool_destroy_work_item shall call sm_exec_begin ]*/
-        SM_RESULT sm_result = sm_exec_begin(threadpool_ptr->sm);
-        if (sm_result != SM_EXEC_GRANTED)
+        
+        /* Codes_SRS_THREADPOOL_LINUX_05_033: [ threadpool_destroy_work_item shall wait for all pending work items to finish execution. ]*/
+        if (InterlockedHL_WaitForValue(&threadpool_work_item->pending_work_item_count, 0, UINT32_MAX) == INTERLOCKED_HL_OK)
```

Comment:
>pending_work_item_count [](http://example.com/codeflow?start=62&length=23)

Overall I think using this pending_work_item_count is wrong. Because it tried to act as a reference count, but it is very bad at it.

For example, assume this:

A work item is created
A thread 1 schedules the work item to execute => pending_work_item_count = 1
Callback executes, => pending_work_item_count = 0
Thread 1 calls destroy_work_item, succesfully executes ln 862.
Thread 2 comes and schedules the work item => pending_work_item_count = 1
Thread 1 frees the work item.

There is no safety against this kind of race.

I believe the correct way to handle this is proper ref counting (either sm or THANDLE).

Note that the previous API we had (which would alloc/schedule/free) does not really have this problem because the lifetime of the work item is handled inside the threadpool.

I think the new API should have the THANDLE(WORK_ITEM) done before we propagate it everywhere. Otherwise, I sense that we would have to revisit again. I think this will remove these race issues we are discussing about.

I understand how changing to a THANDLE needs to change the header of threadpool and the win32 and linux implementation, so I would assume you want to do the change to THANDLE as a separate task - which means that the code here will change again anyway.

Reply:
Hi Dan, yes, it'll be changed later in other task.